File: npro/npro/report/closed_candidate_sourcing/closed_candidate_sourcing.js
// Copyright (c) 2016, GreyCube Technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Closed Candidate Sourcing"] = {
	"filters": [
		{
			fieldname: "from_date",
			label: __("From Date (Job Creation Date)"),
			fieldtype: "Date",
			default: moment().startOf("year").format(),
		},
		{
			fieldname: "to_date",
			label: __("To Date (Job Creation Date)"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
		},
		{
			fieldname: "timespan",
			label: __("Job Creation Date in "),
			fieldtype: "Select",
			options: npro.utils.TIMESPAN_OPTIONS,
			on_change: function (query_report) {
				let date_range = npro.utils.get_date_range(
					query_report.get_values().timespan
				);
				frappe.query_report.set_filter_value({
					from_date: date_range[0],
					to_date: date_range[1],
				});
			},
			default: "This Year",
		},
	]
};

File: npro/npro/report/closed_candidate_sourcing/closed_candidate_sourcing.py
# Copyright (c) 2013, GreyCube Technologies and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe.utils import getdate
import pandas


def execute(filters=None):
    return get_columns(filters), get_data(filters)


def get_data(filters):
    data = frappe.db.sql(
        """ 
        select 
            name, npro_sourcing_owner_cf, customer_cf, job_title, 
            location_cf, contract_duration_cf 
        from 
            `tabJob Opening` tjo 
        {where_conditions}
    """.format(
            where_conditions=get_conditions(filters)
        ),
        filters,
        as_dict=True,
        # debug=True,
    )
    return data


def get_columns(filters):
    return [
        {
            "label": "Job Opening",
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Job Opening",
            "width": 145,
        },
        {
            "label": "Owner",
            "fieldname": "npro_sourcing_owner_cf",
            "fieldtype": "Link",
            "options": "User",
            "width": 145,
        },
        {
            "label": "Customer",
            "fieldname": "customer_cf",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 145,
        },
        {
            "label": "Job Title",
            "fieldname": "job_title",
            "fieldtype": "Data",
            "width": 145,
        },
        {
            "label": "Location",
            "fieldname": "location_cf",
            "width": 145,
        },
        {
            "label": "Contract Duration",
            "fieldname": "contract_duration_cf",
            "fieldtype": "Int",
            "width": 145,
        },
    ]


def get_conditions(filters):
    conditions = ["tjo.status = 'Closed'"]
    if filters.get("from_date"):
        conditions += ["tjo.creation >= %(from_date)s"]
    if filters.get("to_date"):
        conditions += ["tjo.creation <= %(to_date)s"]

    return conditions and " where " + " and ".join(conditions) or ""
